src/test/java/com/cloudbees/jenkins/GitHubPushHookReceiverTest.java

```diff
@@ -0,0 +1,103 @@
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+/**
+ * Test case covering the code responsible to receive the push notification
+ * from GitHub.
+ */
+public class GitHubPushHookReceiverTest extends Assert {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    @Inject
+    GitHubWebHook receiver;
+
+    @Test
+    public void receivePushHookOnFreeStyle() throws IOException, InterruptedException, ExecutionException {
+        j.jenkins.getInjector().injectMembers(this);
+        FreeStyleProject job = j.jenkins.createProject(FreeStyleProject.class, "Test");
+        GitSCM scm = new GitSCM(
+            Collections.singletonList(new UserRemoteConfig("https://github.com/amuniz/github-plugin.git", null, null, "github.com")),
+            Collections.singletonList(new BranchSpec("")),
+            false, Collections.<SubmoduleConfig>emptyList(),
+            null, null, null);
+
+        GitHubPushTrigger trigger = new GitHubPushTrigger();
+        trigger.start(job, false);
+        job.addTrigger(trigger);
+        job.setScm(scm);
+
+        String payload = IOUtils.toString(getClass().getResourceAsStream("/com/cloudbees/jenkins/plugins/github/payload2.json"));
+        receiver.processGitHubPayload(payload, GitHubPushTrigger.class);
+        Run build = waitForBuild(1, job);
+        assertTrue("Build was triggered but did not success: " + build.getLog(300), Result.SUCCESS.equals(build.getResult()));
+    }
+
+    @Test
+    public void receivePushHookOnWorkflow() throws Exception {
+        j.jenkins.getInjector().injectMembers(this);
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "Test Workflow");
+
+        GitHubPushTrigger trigger = new GitHubPushTrigger();
+        trigger.start(job, false);
+        job.addTrigger(trigger);
+        job.setDefinition(new CpsFlowDefinition(
+            "node {\n" +
+            "    git 'https://github.com/amuniz/github-plugin.git'\n" +
+            "}"));
+
+        // Trigger the build once to register SCMs
+        WorkflowRun lastRun = j.assertBuildStatusSuccess(job.scheduleBuild2(0));
+        // Testing hack! This will make the polling believe that there was remote changes to build
+        lastRun.getActions(BuildData.class).get(0).buildsByBranchName = new HashMap<String, Build>();
+
+        // Then simulate a GitHub push
+        String payload = IOUtils.toString(getClass().getResourceAsStream("/com/cloudbees/jenkins/plugins/github/payload2.json"));
+        receiver.processGitHubPayload(payload, GitHubPushTrigger.class);
+        Run build = waitForBuild(2, job);
+        assertTrue("Build was triggered but did not success: " + build.getLog(300), Result.SUCCESS.equals(build.getResult()));
+    }
+
+    private Run waitForBuild(int n, Job job) throws InterruptedException {
+        System.out.println("Waiting for build #" + n + " to appear and finish");
+        int counter = 0;
+        while (true) {
```

Comment:
`while(true)` is bad practice - why not just check counter in while loop and use `fail()` after that?